src/Core/Framework/Struct/Collection.php

```
@@ -81,77 +82,85 @@ public function count(): int
     }
 
     /**
-     * @return list<array-key>
+     * @return list<TKey>
      */
     public function getKeys(): array
     {
         return array_keys($this->elements);
     }
 
     /**
-     * @param array-key $key
+     * @param TKey $key
      */
     public function has($key): bool
     {
         return \array_key_exists($key, $this->elements);
     }
 
     /**
-     * @template T of TElement
+     * @template R of mixed
      *
-     * @param \Closure(T): mixed $closure
+     * @param \Closure(TElement): R $closure
      *
-     * @return array<array-key, mixed>
+     * @return array<TKey, R>
      */
     public function map(\Closure $closure): array
     {
         return array_map($closure, $this->elements);
     }
 
     /**
-     * @template T of TElement
+     * @template R of mixed
      *
-     * @param \Closure(mixed|null, T): mixed $closure
-     * @param mixed|null $initial
+     * @param \Closure(R, TElement): R $closure
+     * @param R $initial
      *
-     * @return mixed|null
+     * @return R
      */
     public function reduce(\Closure $closure, $initial = null)
     {
         return array_reduce($this->elements, $closure, $initial);
     }
 
     /**
-     * @template T of TElement
+     * @template R of mixed
      *
-     * @param \Closure(T): mixed $closure
+     * @param \Closure(TElement): R $closure
      *
-     * @return array<array-key, mixed>
+     * @return array<TKey, R>
      */
     public function fmap(\Closure $closure): array
     {
         return array_filter($this->map($closure));
     }
 
+    /**
+     * @param \Closure(TElement, TElement): int $closure
+     */
     public function sort(\Closure $closure): void
     {
         uasort($this->elements, $closure);
     }
 
     /**
-     * @param class-string $class
+     * @template T
+     *
+     * @param class-string<T> $class
+     *
+     * @return static<TElement&T, TKey>
      */
     public function filterInstance(string $class): static
```

would you mind checking if this works? e.g. with removing this ignore error? 

https://github.com/shopware/shopware/blob/trunk/src/Core/Content/Rule/DataAbstractionLayer/RuleAreaUpdater.php#L291

---

there are also other places, where we do a `@var` annotation after this call. if you have done it correctly, they should all not be necessary anymore :grin:

---

Works as expected :+1: I fear there will be a lot of issues to solve :smile: